Allow for callable classes for resolvers in directives

carlalexander opened the pull request:
This is the PR that fixes #882. 

Instead of adding the `__invoke` method in the `getMethodArgumentParts` method, I moved it to the `Utils` class. I made `methodName` nullable and, if it's `null`, I set it to `__invoke`. We then check if the method name exists.

spawnia commented on src/Support/Utils.php:
```diff
@@ -48,8 +48,12 @@ public static function namespaceClassname(string $classCandidate, array $namespa
-    public static function constructResolver(string $className, string $methodName): Closure
+    public static function constructResolver(string $className, ?string $methodName): Closure
     {
+        if (null === $methodName) {
```
Can we just use a default argument here?

carlalexander replied:
Not right now, no. The `getMethodArgumentPart` method returns an array which you expand right away. So everywhere `methodName` will be `null` when passed to the `constructResolver` method. 

We'd have to rework all the code around that to be able to use a default argument.

carlalexander replied:
This change wasn't working and wasn't relevant in the end. The original change worked fine for detecting errors.

spawnia commented:
The tests are failing.

Can you please stick to the PR template and make sure tests, docs and the changelog are updated?

carlalexander commented:
Sorry about that. Working on tests right now. 😃

carlalexander commented:
I've reworked the PR and added/reworked some of the tests. Do you want all the directives that use resolvers to have tests for it?

What do you want the new text in the docs to be instead of this:
>Consists of two parts: a class name and a method name, separated by an `@` symbol.

spawnia commented on src/Schema/Directives/BaseDirective.php:
```diff
@@ -181,15 +181,18 @@ protected function getMethodArgumentParts(string $argumentName): array
         }
 
+        if (empty($argumentParts[1])) {
+            $argumentParts[1] = '__invoke';
```
Nicely done!

spawnia commented:
I worked through the docs and changed the necessary parts.

Actually, i like this style so much i decided to expand on it a bit and make it an option for any kind of default resolver location. I want to keep compatibility for now, but in `v5` we can make it the new default.

carlalexander commented:
Awesome! Glad to help 😄
